auth/role_collection_access.go

```diff
@@ -0,0 +1,225 @@
+	if base.IsDefaultCollection(scope, collection) {
+		role.SetChannelInvalSeq(invalSeq)
+		return
+	}
+
+	cc := role.getOrCreateCollectionAccess(scope, collection)
+	cc.ChannelInvalSeq = invalSeq
+}
+
+func (role *roleImpl) collectionInvalidatedChannels(scope, collection string) ch.TimedSet {
+	if base.IsDefaultCollection(scope, collection) {
+		return role.InvalidatedChannels()
+	}
+
+	if cc, ok := role.getCollectionAccess(scope, collection); ok {
+		if cc.ChannelInvalSeq != 0 {
+			return cc.Channels_
+		}
+	}
+	return nil
+}
+
+func (role *roleImpl) CollectionChannelHistory(scope, collection string) TimedSetHistory {
+	if base.IsDefaultCollection(scope, collection) {
+		return role.ChannelHistory()
+	}
+
+	if cc, ok := role.getCollectionAccess(scope, collection); ok {
+		return cc.ChannelHistory_
+	}
+	return nil
+}
+
+func (role *roleImpl) SetCollectionChannelHistory(scope, collection string, history TimedSetHistory) {
+	if base.IsDefaultCollection(scope, collection) {
+		role.SetChannelHistory(history)
+		return
+	}
+
+	cc := role.getOrCreateCollectionAccess(scope, collection)
+	cc.ChannelHistory_ = history
+}
+
+// Returns true if the Role is allowed to access the channel.
+// A nil Role means access control is disabled, so the function will return true.
+func (role *roleImpl) CanSeeCollectionChannel(scope, collection, channel string) bool {
+	if base.IsDefaultCollection(scope, collection) {
+		return role.CanSeeChannel(channel)
+	}
+
+	if role == nil {
+		return true
+	}
+	if cc, ok := role.getCollectionAccess(scope, collection); ok {
+		return cc.Channels().Contains(channel) || cc.Channels().Contains(ch.UserStarChannel)
+	}
+	return false
+}
+
+// Returns the sequence number since which the Role has been able to access the channel, else zero.
+func (role *roleImpl) canSeeCollectionChannelSince(scope, collection, channel string) uint64 {
+	if base.IsDefaultCollection(scope, collection) {
+		return role.CanSeeChannelSince(channel)
+	}
+
+	if cc, ok := role.getCollectionAccess(scope, collection); ok {
+		seq := cc.Channels()[channel]
+		if seq.Sequence == 0 {
+			seq = cc.Channels()[ch.UserStarChannel]
+		}
+	}
+	return 0
+}
+
+func (role *roleImpl) authorizeAllCollectionChannels(scope, collection string, channels base.Set) error {
+	if base.IsDefaultCollection(scope, collection) {
+		return role.AuthorizeAllChannels(channels)
+	}
+
+	if ca, ok := role.getCollectionAccess(scope, collection); ok {
+		var forbidden []string
+		for channel := range channels {
+			if !ca.CanSeeChannel(channel) {
+				if forbidden == nil {
+					forbidden = make([]string, 0, len(channels))
+				}
+				forbidden = append(forbidden, channel)
+			}
+		}
+		if forbidden != nil {
+			return role.UnauthError(fmt.Sprintf("You are not allowed to see channels %v", forbidden))
+		}
+		return nil
+	}
+	return role.UnauthError(fmt.Sprintf("Unauthorized to see channels %v", channels))
```

Comment:
Do you want this error message or say no such scope and collection? is that leaking information?